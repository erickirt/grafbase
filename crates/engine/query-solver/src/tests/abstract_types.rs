use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
    AGENCY @join__graph(name: "agency", url: "http://localhost:4200/abstract-types/agency")
    BOOKS @join__graph(name: "books", url: "http://localhost:4200/abstract-types/books")
    INVENTORY @join__graph(name: "inventory", url: "http://localhost:4200/abstract-types/inventory")
    MAGAZINES @join__graph(name: "magazines", url: "http://localhost:4200/abstract-types/magazines")
    PRODUCTS @join__graph(name: "products", url: "http://localhost:4200/abstract-types/products")
    REVIEWS @join__graph(name: "reviews", url: "http://localhost:4200/abstract-types/reviews")
    USERS @join__graph(name: "users", url: "http://localhost:4200/abstract-types/users")
}

type Agency
    @join__type(graph: AGENCY, key: "id")
    @join__type(graph: PRODUCTS)
{
    id: ID!
    companyName: String @join__field(graph: AGENCY)
}

type Book implements Product & Similar
    @join__implements(graph: INVENTORY, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Similar")
    @join__implements(graph: REVIEWS, interface: "Product")
    @join__implements(graph: REVIEWS, interface: "Similar")
    @join__type(graph: BOOKS, key: "id")
    @join__type(graph: INVENTORY, key: "id")
    @join__type(graph: PRODUCTS, key: "id")
    @join__type(graph: REVIEWS, key: "id")
{
    id: ID!
    title: String @join__field(graph: BOOKS)
    dimensions: ProductDimension @join__field(graph: PRODUCTS)
    delivery(zip: String): DeliveryEstimates @join__field(graph: INVENTORY, requires: "dimensions { size weight }")
    sku: String @join__field(graph: PRODUCTS)
    createdBy: User @join__field(graph: PRODUCTS)
    similar: [Book] @join__field(graph: PRODUCTS)
    hidden: Boolean @join__field(graph: PRODUCTS)
    publisherType: PublisherType @join__field(graph: PRODUCTS)
    reviewsCount: Int! @join__field(graph: REVIEWS)
    reviewsScore: Float! @join__field(graph: REVIEWS)
    reviews: [Review!]! @join__field(graph: REVIEWS)
    reviewsOfSimilar: [Review!]! @join__field(graph: REVIEWS, requires: "similar { id }")
}

type DeliveryEstimates
    @join__type(graph: INVENTORY)
{
    estimatedDelivery: String
    fastestDelivery: String
}

type Group
    @join__type(graph: AGENCY, key: "id")
{
    id: ID!
    name: String
}

type Magazine implements Product & Similar
    @join__implements(graph: INVENTORY, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Similar")
    @join__implements(graph: REVIEWS, interface: "Product")
    @join__implements(graph: REVIEWS, interface: "Similar")
    @join__type(graph: INVENTORY, key: "id")
    @join__type(graph: MAGAZINES, key: "id")
    @join__type(graph: PRODUCTS, key: "id")
    @join__type(graph: REVIEWS, key: "id")
{
    id: ID!
    dimensions: ProductDimension @join__field(graph: PRODUCTS)
    delivery(zip: String): DeliveryEstimates @join__field(graph: INVENTORY, requires: "dimensions { size weight }")
    title: String @join__field(graph: MAGAZINES)
    sku: String @join__field(graph: PRODUCTS)
    createdBy: User @join__field(graph: PRODUCTS)
    similar: [Magazine] @join__field(graph: PRODUCTS)
    hidden: Boolean @join__field(graph: PRODUCTS)
    publisherType: PublisherType @join__field(graph: PRODUCTS)
    reviewsCount: Int! @join__field(graph: REVIEWS)
    reviewsScore: Float! @join__field(graph: REVIEWS)
    reviews: [Review!]! @join__field(graph: REVIEWS)
    reviewsOfSimilar: [Review!]! @join__field(graph: REVIEWS, requires: "similar { id }")
}

type ProductDimension
    @join__type(graph: INVENTORY)
    @join__type(graph: PRODUCTS)
{
    size: String
    weight: Float
}

type Query
    @join__type(graph: AGENCY)
    @join__type(graph: BOOKS)
    @join__type(graph: INVENTORY)
    @join__type(graph: MAGAZINES)
    @join__type(graph: PRODUCTS)
    @join__type(graph: REVIEWS)
    @join__type(graph: USERS)
{
    books: [Book] @join__field(graph: BOOKS)
    magazines: [Magazine] @join__field(graph: MAGAZINES)
    products: [Product] @join__field(graph: PRODUCTS)
    similar(id: ID!): [Product] @join__field(graph: PRODUCTS)
    review(id: Int!): Review @join__field(graph: REVIEWS)
}

type Review
    @join__type(graph: REVIEWS)
{
    id: Int!
    body: String!
    product: Product
}

type Self
    @join__type(graph: PRODUCTS)
{
    email: String
}

type User
    @join__type(graph: PRODUCTS, key: "email")
    @join__type(graph: USERS, key: "email")
{
    email: ID!
    totalProductsCreated: Int
    name: String @join__field(graph: USERS)
}

interface Product
    @join__type(graph: INVENTORY)
    @join__type(graph: PRODUCTS)
    @join__type(graph: REVIEWS)
{
    id: ID!
    dimensions: ProductDimension @join__field(graph: INVENTORY) @join__field(graph: PRODUCTS)
    delivery(zip: String): DeliveryEstimates @join__field(graph: INVENTORY)
    sku: String @join__field(graph: PRODUCTS)
    createdBy: User @join__field(graph: PRODUCTS)
    hidden: Boolean @inaccessible @join__field(graph: PRODUCTS)
    reviewsCount: Int! @join__field(graph: REVIEWS)
    reviewsScore: Float! @join__field(graph: REVIEWS)
    reviews: [Review!]! @join__field(graph: REVIEWS)
}

interface Similar
    @join__type(graph: PRODUCTS)
    @join__type(graph: REVIEWS)
{
    similar: [Product]
}

union PublisherType
    @join__type(graph: AGENCY)
    @join__type(graph: PRODUCTS)
    @join__unionMember(graph: AGENCY, member: "Agency")
    @join__unionMember(graph: PRODUCTS, member: "Agency")
    @join__unionMember(graph: AGENCY, member: "Group")
    @join__unionMember(graph: PRODUCTS, member: "Self")
 = Agency | Group | Self
"#;

#[test]
fn interface_field_provided_by_implementors() {
    // delivery is not accessible on the Similar interface, but it is on all of its implementors.
    assert_solving_snapshots!(
        "interface_field_provided_by_implementors",
        SCHEMA,
        r#"
        query {
          similar(id: "p1") {
            id
            sku
            delivery(zip: "1234") {
              fastestDelivery
              estimatedDelivery
            }
          }
        }
        "#
    );
}

#[test]
fn nested_interface_field_provided_by_implementors() {
    assert_solving_snapshots!(
        "nested_interface_field_provided_by_implementors",
        SCHEMA,
        r#"
        {
          products {
            id
            reviews {
              product {
                sku
                ... on Magazine {
                  title
                }
                ... on Book {
                  reviewsCount
                }
              }
            }
          }
        }
        "#
    );
}

#[test]
fn unreachable_object() {
    // Group is never reachable.
    assert_solving_snapshots!(
        "unreachable_object",
        SCHEMA,
        r#"
        query {
          products {
            id
            ... on Magazine {
              publisherType {
                ...Publisher
              }
            }
            ... on Book {
              publisherType {
                ...Publisher
              }
            }
          }
        }

        fragment Publisher on PublisherType {
          ... on Agency {
            id
            companyName
          }
          ... on Self {
            email
          }
          ... on Group {
            name
          }
        }
        "#
    );
}
