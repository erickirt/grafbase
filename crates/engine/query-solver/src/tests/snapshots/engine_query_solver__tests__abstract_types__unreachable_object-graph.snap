---
source: crates/engine/query-solver/src/tests/abstract_types.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Query.products\" ]\n    2 [ label = \"Product.id\" ]\n    3 [ label = \"Magazine.publisherType\" ]\n    4 [ label = \"Book.publisherType\" ]\n    5 [ label = \"Agency.id\" ]\n    6 [ label = \"Agency.companyName\" ]\n    7 [ label = \"Self.email\" ]\n    9 [ label = \"Agency.id\" ]\n    10 [ label = \"Agency.companyName\" ]\n    11 [ label = \"Self.email\" ]\n    13 [ label = \"Root#products\", shape=parallelogram, color=dodgerblue ]\n    14 [ label = \"products#products\", shape=box, color=dodgerblue ]\n    15 [ label = \"id#products\", shape=box, color=dodgerblue ]\n    16 [ label = \"publisherType#products\", shape=box, color=dodgerblue ]\n    17 [ label = \"id#products\", shape=box, color=dodgerblue ]\n    18 [ label = \"FedEntity#agency\", shape=parallelogram, color=dodgerblue ]\n    19 [ label = \"companyName#agency\", shape=box, color=dodgerblue ]\n    20 [ label = \"email#products\", shape=box, color=dodgerblue ]\n    21 [ label = \"publisherType#products\", shape=box, color=dodgerblue ]\n    22 [ label = \"id#products\", shape=box, color=dodgerblue ]\n    23 [ label = \"FedEntity#agency\", shape=parallelogram, color=dodgerblue ]\n    24 [ label = \"companyName#agency\", shape=box, color=dodgerblue ]\n    25 [ label = \"email#products\", shape=box, color=dodgerblue ]\n    0 -> 1 [  ]\n    1 -> 2 [  ]\n    1 -> 3 [  ]\n    1 -> 4 [  ]\n    3 -> 5 [  ]\n    3 -> 6 [  ]\n    3 -> 7 [  ]\n    4 -> 9 [  ]\n    4 -> 10 [  ]\n    4 -> 11 [  ]\n    0 -> 13 [ color=royalblue,fontcolor=royalblue ]\n    0 -> 13 [ style=dashed,arrowhead=none ]\n    13 -> 14 [ color=royalblue,fontcolor=royalblue ]\n    14 -> 1 [ color=violet,arrowhead=none ]\n    14 -> 15 [ color=royalblue,fontcolor=royalblue ]\n    15 -> 2 [ color=violet,arrowhead=none ]\n    14 -> 16 [ color=royalblue,fontcolor=royalblue ]\n    16 -> 3 [ color=violet,arrowhead=none ]\n    16 -> 17 [ color=royalblue,fontcolor=royalblue ]\n    17 -> 5 [ color=violet,arrowhead=none ]\n    16 -> 18 [ color=royalblue,fontcolor=royalblue ]\n    3 -> 18 [ style=dashed,arrowhead=none ]\n    18 -> 19 [ color=royalblue,fontcolor=royalblue ]\n    19 -> 6 [ color=violet,arrowhead=none ]\n    16 -> 20 [ color=royalblue,fontcolor=royalblue ]\n    20 -> 7 [ color=violet,arrowhead=none ]\n    14 -> 21 [ color=royalblue,fontcolor=royalblue ]\n    21 -> 4 [ color=violet,arrowhead=none ]\n    21 -> 22 [ color=royalblue,fontcolor=royalblue ]\n    22 -> 9 [ color=violet,arrowhead=none ]\n    21 -> 23 [ color=royalblue,fontcolor=royalblue ]\n    4 -> 23 [ style=dashed,arrowhead=none ]\n    23 -> 24 [ color=royalblue,fontcolor=royalblue ]\n    24 -> 10 [ color=violet,arrowhead=none ]\n    21 -> 25 [ color=royalblue,fontcolor=royalblue ]\n    25 -> 11 [ color=violet,arrowhead=none ]\n    23 -> 9 [ color=orangered,arrowhead=inv ]\n    18 -> 5 [ color=orangered,arrowhead=inv ]\n}\n"
---
digraph {
    0 [ root]
    1 [ Query.products]
    2 [ Product.id]
    3 [ Magazine.publisherType]
    4 [ Book.publisherType]
    5 [ Agency.id]
    6 [ Agency.companyName]
    7 [ Self.email]
    9 [ Agency.id]
    10 [ Agency.companyName]
    11 [ Self.email]
    13 [ Root#products]
    14 [ products#products]
    15 [ id#products]
    16 [ publisherType#products]
    17 [ id#products]
    18 [ FedEntity#agency]
    19 [ companyName#agency]
    20 [ email#products]
    21 [ publisherType#products]
    22 [ id#products]
    23 [ FedEntity#agency]
    24 [ companyName#agency]
    25 [ email#products]
    0 -> 1 [ label = "Field" ]
    1 -> 2 [ label = "Field" ]
    1 -> 3 [ label = "Field" ]
    1 -> 4 [ label = "Field" ]
    3 -> 5 [ label = "Field" ]
    3 -> 6 [ label = "Field" ]
    3 -> 7 [ label = "Field" ]
    4 -> 9 [ label = "Field" ]
    4 -> 10 [ label = "Field" ]
    4 -> 11 [ label = "Field" ]
    0 -> 13 [ label = "CreateChildResolver" ]
    0 -> 13 [ label = "HasChildResolver" ]
    13 -> 14 [ label = "CanProvide" ]
    14 -> 1 [ label = "Provides" ]
    14 -> 15 [ label = "CanProvide" ]
    15 -> 2 [ label = "Provides" ]
    14 -> 16 [ label = "CanProvide" ]
    16 -> 3 [ label = "Provides" ]
    16 -> 17 [ label = "CanProvide" ]
    17 -> 5 [ label = "Provides" ]
    16 -> 18 [ label = "CreateChildResolver" ]
    3 -> 18 [ label = "HasChildResolver" ]
    18 -> 19 [ label = "CanProvide" ]
    19 -> 6 [ label = "Provides" ]
    16 -> 20 [ label = "CanProvide" ]
    20 -> 7 [ label = "Provides" ]
    14 -> 21 [ label = "CanProvide" ]
    21 -> 4 [ label = "Provides" ]
    21 -> 22 [ label = "CanProvide" ]
    22 -> 9 [ label = "Provides" ]
    21 -> 23 [ label = "CreateChildResolver" ]
    4 -> 23 [ label = "HasChildResolver" ]
    23 -> 24 [ label = "CanProvide" ]
    24 -> 10 [ label = "Provides" ]
    21 -> 25 [ label = "CanProvide" ]
    25 -> 11 [ label = "Provides" ]
    23 -> 9 [ label = "Requires" ]
    18 -> 5 [ label = "Requires" ]
}
